Remove slimmer/static dependency

KludgeKML opened the pull request:
Removes the need to call out to static via slimmer for layouts.

- Remove slimmer
- Layouts now directly call the layout_for_public component
- Emergency/Global banner support provided by the govuk_web_banners gem

https://trello.com/c/O6JhukLu/245-remove-slimmer-email-alert-frontend, [Jira issue PNP-7340](https://gov-uk.atlassian.net/browse/PNP-7340)

Practical differences in rendered pages:
- stylesheets are now more granular (the static application stylesheet is no longer downloaded. The application stylesheet contains the styles used in all pages, bundled for compression efficiency, and any styles only used in one layout are included as individual component sheets). This has both pros and cons (pros: more granular/lightweight for email-alert-frontend pages, cons: less efficient for overall gov.uk browsing. Cons will be rectified as we move towards shared component stylesheet folders).
- pages now no longer have the meta `http-equiv="Content-Type"` element, as that is now obsolete (`<meta charset="utf-8">`) covers it.
- slimmer used to hoist a lot of meta tags to the top of the head. They now appear in the order specified in the layout.

⚠️  This repo is Continuously Deployed: make sure you [follow the guidance](https://docs.publishing.service.gov.uk/manual/development-pipeline.html#merge-your-own-pull-request) ⚠️

KludgeKML committed:
Remove Slimmer Gem and config

- TODO about choosing account layout will be addressed
  in a later commit, for the moment just remove all slimmer
  calls. We intend to replace this with a custom layout file
  for accounts.

KludgeKML committed:
Add govuk_web_banners gem

- Add banner JS dependencies
- Add global/emergency banner support in layouts (no
  global banner for accounts, see: https://github.com/alphagov/static/blob/fb7e8c9ba9f49a20f63a3e001cd12982477added/app/views/root/gem_layout_account_manager.html.erb#L9)
- Add initializer for emergency banner REDIS support
- Mock redis client in all tests (we're not explicitly
  testing emergency banner support here, since we assume
  that's tested adequately in the gem).

MartinJJones reviewed (commented):
Nice work!, I've had an initial look through the changes and made a few suggestions.

I would be interested in taking a further look at the impact on performance once the other CSS/JS files have been added. For example, loading all stylesheets individually compared to the current approach with static. Happy to catch up / pair on this further if that works for you 👍

KludgeKML commented on app/assets/javascripts/application.js:
```diff
@@ -1,3 +1,9 @@
 //= require govuk_publishing_components/components/error-summary
 //= require govuk_publishing_components/components/radio
+//= require govuk_publishing_components/components/feedback
+//= require govuk_publishing_components/components/global-banner
```
Ah, good spot. Remove!

KludgeKML commented on app/views/layouts/account.html.erb:
```diff
@@ -0,0 +1,76 @@
+  title: yield(:title),
+  blue_bar: false,
+  emergency_banner: render("govuk_web_banners/emergency_banner"),
+  global_banner: render("govuk_web_banners/global_banner"),
```
No, good spot. Removed!

KludgeKML commented on app/assets/javascripts/application.js:
```diff
@@ -1,3 +1,9 @@
 //= require govuk_publishing_components/components/radio
+//= require govuk_publishing_components/components/feedback
+//= require govuk_publishing_components/components/global-banner
+//= require govuk_publishing_components/components/layout-super-navigation-header
```
Fixed!

KludgeKML commented on app/views/layouts/application.html.erb:
```diff
@@ -1,47 +1,38 @@
+  <% add_gem_component_stylesheet("layout-footer") %>
+  <% add_gem_component_stylesheet("layout-for-public") %>
+  <% add_gem_component_stylesheet("layout-super-navigation-header") %>
+  <%= render_component_stylesheets %>
```
Fixed!

KludgeKML committed:
Add account-specific layout and code to support

- this replicates/replaces the code from:
  https://github.com/alphagov/static/blob/f88f6e405c63e384411bdc7f25ec926b63d29f06/app/views/root/gem_layout_account_manager.html.erb
  https://github.com/alphagov/static/blob/f88f6e405c63e384411bdc7f25ec926b63d29f06/app/views/root/_gem_base.html.erb

  ...and moves this configuration to the only app that uses it.

- add rendering-application meta tag (previously supplied by slimmer)

fix: account

fix: account

andysellick commented on app/assets/stylesheets/application.scss:
```diff
@@ -2,3 +2,5 @@
 
 // https://github.com/alphagov/govuk_publishing_components/blob/main/docs/set-up-individual-component-css-loading.md
+@import "govuk_publishing_components/govuk_frontend_support";
+@import "govuk_publishing_components/component_support";
```
Static includes a number of component stylesheets needed by all applications e.g. heading - do some of those need adding here?

KludgeKML replied:
We've added here a bundled set of components that are used by all layouts in the app. The individual stylesheet for headings will be brought in as an individual component.

andysellick commented on app/views/layouts/application.html.erb:
```diff
@@ -1,47 +1,42 @@
+  <div class="govuk-width-container" id="wrapper">
+    <%= yield :back_link %>
+    <main class="govuk-main-wrapper<%= " govuk-main-wrapper--l" if yield(:back_link).blank?%>" role="main" id="content">
+      <div id="email-alert-frontend">
```
Is this div needed? I can't see any reference to it elsewhere in the code.

KludgeKML replied:
Good question. It's in the existing code, but removing it seems to make no difference, and as you say there isn't any CSS or JS referring to it. I'll take it out.

KludgeKML replied:
(Done)

andysellick commented on app/views/layouts/application.html.erb:
```diff
@@ -1,47 +1,42 @@
+  <%= csp_meta_tag %>
+  <meta name="robots" content="noindex, nofollow">
+
+  <% add_gem_component_stylesheet("cookie-banner") %>
```
Just to check - these are all the stylesheets that are needed prior to something loading, which is why they need to be hoisted here, and not included in application.scss? Is that right?

KludgeKML replied:
application.scss should include stylesheets needed by all pages, but these stylesheets are ones only needed by these specific layouts, so they're hoisted above the call to layout_for_public (in fact there's now nothing in this specific layout, but the account layout has the cross-service-header and phase-banner stylesheets included, since they're needed for anything using that layout, but not for anything using the normal layout).

andysellick commented on app/views/subscription_authentication/expired.html.erb:
```diff
@@ -1,5 +1,7 @@
 <% content_for :title, t("subscription_authentication.expired.title") %>
-<% content_for :meta_description, t("subscription_authentication.expired.description") %>
+<% content_for :head do %>%
```
`<% content_for :head do %>%` <-- typo?

KludgeKML replied:
Thanks!

KludgeKML replied:
Fixed

andysellick commented on app/assets/javascripts/application.js:
```diff
@@ -1,3 +1,13 @@
+//= require govuk_publishing_components/components/search-with-autocomplete
+//= require govuk_publishing_components/components/skip-link
+
+//= require govuk_web_banners/dependencies
```
The last commit says `Add JS used by global banners` but doesn't include any JS changes, are those scripts already included with this line?

KludgeKML replied:
Done!

andysellick reviewed (commented):
Looking really good, just a few comments/questions.

MartinJJones commented on app/views/layouts/account.html.erb:
```diff
@@ -0,0 +1,75 @@
+  product_name: "GOV.UK email subscriptions",
+  show_account_layout: true,
+  show_cross_service_header: true,
+  show_explore_header: false,
```
non-blocking: we should be able to delete this line now, the `show_explore_header` option was removed from layout_for_public in [v54.0.0](https://github.com/alphagov/govuk_publishing_components/blob/main/CHANGELOG.md#5400) of govuk_publishing_components

MartinJJones commented on app/views/layouts/application.html.erb:
```diff
@@ -1,47 +1,42 @@
+
+<%= render "govuk_publishing_components/components/layout_for_public", {
+      title: yield(:title),
+      show_explore_header: true,
```
non-blocking: we should be able to delete this line now, the `show_explore_header` option was removed from layout_for_public in [v54.0.0](https://github.com/alphagov/govuk_publishing_components/blob/main/CHANGELOG.md#5400) of govuk_publishing_components

KludgeKML committed:
Add basic elements bundle

- although we could bring in all these elements as individual
  stylesheets, the compression benefit we get by bundling all
  these styles (which are used on all pages in the app)
  together is considerable. In tests the bundled styles led
  to an overall style size when compressed of 20 Kb (compared
  to 34 Kb with all individual styles, or 22 Kb with the
  current static system). There may be further refinements
  we can make, but for the moment this is reasonable.
- We bundle the basic component stylesheets in the application.css
- We also need to add these to the custom_css_exclude_list
  so that the app doesn't duplicate them by bringing them
  in as individual files too.
- Also include the JS files needed for these components
  (previously supplied by static's JS file)

KludgeKML committed:
Add basic layout with direct component calls

- Update layout to use layout_for_public component directly
- we call render_component_stylesheetsas part of the content_for :head,
  which is yielded to in the layout.
- Layout removes the metadata_description section, this will be handled in
  the following commit.

KludgeKML committed:
Update how expired template provides metadata

- This was previously handled as a content_for :meta_description
  in the layout, but it was only used by this template. Since we
  now have a content_for :head in the layout, use it here to
  append to that block before it's sent to the layout_for_public
  component.
- Add a test for this, since it wasn't being checked for and it
  is useful to confirm this refactor.

Audit Trail:
https://github.com/alphagov/email-alert-frontend/blob/bc3ffdcd72f09708635558350493d5ef9549a018/app/views/layouts/application.html.erb#L34-L36
https://github.com/alphagov/email-alert-frontend/blob/bc3ffdcd72f09708635558350493d5ef9549a018/app/views/subscription_authentication/expired.html.erb#L2

MartinJJones reviewed (commented):
Awesome work on this!

I have a couple more comments/suggestions, other than that, I'm happy to approve the changes 👍

KludgeKML committed:
perf: Optimise contents of link header

- By calling stylesheet_link_tag "application" without actually
  rendering it, we ensure that it gets into the link header, allowing
  browsers to begin downloading it early so that it doesn't block
  rendering.

KludgeKML commented on app/assets/stylesheets/application.scss:
```diff
@@ -2,3 +2,17 @@
+
+@import "govuk_publishing_components/components/button";
+@import "govuk_publishing_components/components/cookie-banner";
+@import "govuk_publishing_components/components/feedback";
```
Done!

MartinJJones reviewed (approved):
The changes look good to me, nice work! 👍
